## scRNA-seq

A basic challenge of using droplets for transcriptomics is to retain a molecular memory of the identity of the cell from which each mRNA transcript was isolated. To accomplish this, a molecular barcoding strategy was developed to **remember the cell-of-origin of each mRNA**.

### count matrix

In the count matrix generated from scRNA-seq data, **the rows typically represent individual cells, while the columns represent different genes or transcripts**. Each entry in the matrix represents the number of sequencing reads that align to a specific gene in a particular cell. The count matrix provides information about the expression levels of various genes across individual cells, allowing for the analysis of gene expression patterns, identification of cell types, and exploration of cellular heterogeneity within a population.

## scATAC-seq

### Peak

In scATAC-seq, a "peak" refers to a region of the genome that exhibits higher chromatin accessibility than the surrounding genomic regions. These regions often correspond to regulatory elements such as promoters, enhancers, or other functional elements. Peaks are identified by analyzing the read coverage and distribution across the genome in scATAC-seq data. They provide information about regions of the genome that are more accessible and potentially involved in gene regulation or other biological processes.



